contracts/collections/sg721-base/src/contract.rs

```diff
@@ -225,27 +223,63 @@ where
         let current_royalty_info = collection
             .royalty_info
-            .as_ref()
-            .map(|royalty_info| royalty_info.to_response());
+            .map(|royalty_info| royalty_info.into());
 
         let new_royalty_info = collection_msg
             .royalty_info
             .unwrap_or_else(|| current_royalty_info.clone());
 
+        // get the factory from the minter to get
+        // max_royalty, max_royalty_increase_rate
+        let minter_address = self.parent.minter(deps.as_ref())?.minter.unwrap();
+        let minter_config: MinterConfig<T> = deps
+            .querier
+            .query_wasm_smart(minter_address, &MinterQueryMsg::Config {})?;
+        let factory_address = minter_config.factory;
+        let factory_params: ParamsResponse<T> = deps
+            .querier
+            .query_wasm_smart(factory_address, &Sg2QueryMsg::Params {})?;
+        let max_royalty = factory_params.params.max_royalty_bps.bps_to_decimal();
+        let max_royalty_increase_rate = factory_params
+            .params
+            .max_royalty_increase_rate_bps
+            .bps_to_decimal();
+
         // reminder: collection_msg.royalty_info is Option<Option<RoyaltyInfoResponse>>
-        collection.royalty_info = if let Some(royalty_info) = new_royalty_info {
+        collection.royalty_info = if let Some(new_royalty_info_res) = new_royalty_info {
+            if new_royalty_info_res.share > max_royalty {
+                return Err(ContractError::RoyaltyShareTooHigh {});
+            }
             // update royalty info to equal or less, else throw error
-            if let Some(royalty_info_res) = current_royalty_info {
-                if royalty_info.share > royalty_info_res.share {
-                    return Err(ContractError::RoyaltyShareIncreased {});
+            if let Some(curr_royalty_info_res) = current_royalty_info {
+                if new_royalty_info_res.share
+                    > curr_royalty_info_res.share + max_royalty_increase_rate
+                {
+                    return Err(ContractError::RoyaltyShareIncreasedTooMuch {});
+                }
+
+                // now the royalty increase is within range
+
+                // make sure the update time is after 24 hours
+                let end = (Expiration::AtTime(curr_royalty_info_res.updated_at) + DAY)?;
+                let update_too_soon = !end.is_expired(&env.block);
+
+                // if royalty share changed and current time is after last royalty update + min duration
+                // royalty_updated_at is always Some because it is set in instantiate
+                if (new_royalty_info_res.share != curr_royalty_info_res.share) && update_too_soon {
+                    return Err(ContractError::RoyaltyUpdateTooSoon {});
                 }
             } else {
-                return Err(ContractError::RoyaltyShareIncreased {});
+                return Err(ContractError::RoyaltyShareIncreasedTooMuch {});
             }
```

Comment:
This here doesn't allow to increase it if it was empty before. Perhaps you want to allow it now with the same time interval.

Reply:
If current_royalty is `None`, there is no `RoyaltyInfo.updated_at` to compare